src/graphic/animation/animation.cc

```diff
@@ -0,0 +1,195 @@
+	sound_effect_(kNoSoundEffect),
+	sound_priority_(kFxPriorityLowest) {
+	try {
+		// Sound
+		if (table.has_key("sound_effect")) {
+			std::unique_ptr<LuaTable> sound_effects = table.get_table("sound_effect");
+			sound_effect_ =
+			   SoundHandler::register_fx(SoundType::kAmbient, sound_effects->get_string("path"));
+
+			if (sound_effects->has_key<std::string>("priority")) {
+				sound_priority_ = sound_effects->get_int("priority");
+			}
+
+			if (sound_priority_ < kFxPriorityLowest) {
+				throw Widelands::GameDataError(
+				   "Minmum priority for sounds is %d, but only %d was specified for %s",
+				   kFxPriorityLowest, sound_priority_, sound_effects->get_string("path").c_str());
+			}
+		}
+	} catch (const LuaError& e) {
+		throw wexception("Error in animation table: %s", e.what());
+	}
+	assert(frametime_ > 0);
+}
+
+const Animation::MipMapEntry& Animation::mipmap_entry(float scale) const {
+	assert(mipmaps_.count(scale) == 1);
+	const MipMapEntry& mipmap = *mipmaps_.at(scale);
+	mipmap.ensure_graphics_are_loaded();
+	return mipmap;
+}
+
+Rectf Animation::source_rectangle(const int percent_from_bottom, float scale) const {
+	const MipMapEntry& mipmap = mipmap_entry(find_best_scale(scale));
+	const float h = percent_from_bottom * mipmap.height() / 100;
+	// Using floor for pixel perfect positioning
+	return Rectf(0.f, std::floor(mipmap.height() - h), mipmap.width(), h);
+}
+
+Rectf Animation::destination_rectangle(const Vector2f& position,
+                                                const Rectf& source_rect,
+                                                const float scale) const {
+	const float best_scale = find_best_scale(scale);
+	return Rectf(position.x - (hotspot_.x - source_rect.x / best_scale) * scale,
+	             position.y - (hotspot_.y - source_rect.y / best_scale) * scale,
+	             source_rect.w * scale / best_scale, source_rect.h * scale / best_scale);
+}
+
+uint16_t Animation::nr_frames() const {
+	assert(nr_frames_ > 0);
+	return nr_frames_;
+}
+
+int Animation::height() const {
+	return mipmap_entry(1.0f).height();
+}
+
+int Animation::width() const {
+	return mipmap_entry(1.0f).width();
+}
+
+uint32_t Animation::frametime() const {
+	assert(frametime_ > 0);
+	return frametime_;
+}
+
+const Vector2i& Animation::hotspot() const {
+	return hotspot_;
+}
+
+uint32_t Animation::current_frame(uint32_t time) const {
+	if (nr_frames() > 1) {
+		return (play_once_ && time / frametime_ > static_cast<uint32_t>(nr_frames() - 1)) ?
+		          static_cast<uint32_t>(nr_frames() - 1) :
+		          time / frametime_ % nr_frames();
+	}
+	return 0;
+}
+
+void Animation::add_available_scales(const std::string& basename, const std::string& directory) {
+	for (const auto& scale : kSupportedScales) {
+		add_scale_if_files_present(basename, directory, scale.first, scale.second);
+	}
+
+	if (mipmaps_.count(1.0f) == 0) {
+		// There might be only 1 scale
+		add_scale_if_files_present(basename, directory, 1.0f, "");
```

Comment:
Since we can assume that kSupportedScales always contains 1×, is there really a reason why it should work on the second try if it didn´t on the first?

Reply:
This is when the scale got omitted from the filename, so we can handle `idle.png` instead of `idle_1.png` if there is only 1 scale available.